ethcore/src/miner/miner.rs

```diff
@@ -600,6 +605,10 @@ impl Miner {
 							TransactionOrigin::Local | TransactionOrigin::RetractedBlock => {
 								transaction_queue.add(tx, origin, &fetch_account, &gas_required)
 							},
+							TransactionOrigin::External if chain.transaction_block(TransactionID::Hash(tx.hash())).is_some() => {
```

Comment:
I think it would make sense to move the check even higher, before `let origin` to skip ec recovery too.
Also the error should probably be returned for local transactions as well.